core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableThetaJoin.java

```diff
@@ -45,6 +46,7 @@
 /** Implementation of {@link org.apache.calcite.rel.core.Join} in
  * {@link org.apache.calcite.adapter.enumerable.EnumerableConvention enumerable calling convention}
  * that allows conditions that are not just {@code =} (equals). */
+@Deprecated // to be removed before 2.0
```

Comment:
I do not understand why we should deprecate EnumerableThetaJoin: it is the only implementation in enumerable convention that can implement any type of join condition

Reply:
Theta join can be implement through other like join, e.g. NestedLoopJoin, i want to refactor it also, but not in this patch, i agree to remove the annotation now

Reply:
I don't see why we need theta join, isn't it just a nestedloop join?

Reply:
Actually, looking at BuiltInMethod.THETA_JOIN, I agree it is the pure definition of a NestedLoopJoin (without the correlating variable mechanism). So, it would seem that we two implementations of nested loop join (EnumerableCorrelate and EnumerableThetaJoin). What if we did the following:
- Rename EnumerableThetaJoin as EnumerableNestedLoopJoin: right now this is most powerful (albeit most likely inefficient) way to implement any type of join: inner, left, right, outer; with equi or non-equi condition (although the inner and equi ones actually go through EnumerableJoin for efficiency). And I think EnumerableThetaJoin will allow us to implement non-equi semijoins too.
- Keep EnumerableCorrelate as it is, do not rename it. I agree that it is also a nested loop implementation, but it uses the correlating variable approach, so I think keeping its current name will keep things clearer.
This is an alternative approach that differs from the original idea, I'm not sure if it's the right one, but me be worth to discuss it,

Reply:
I agree to rename EnumerableThetaJoin as EnumerableNestedLoopJoin and enhance the implementation of the current EnumerableNestedLoopJoin, i would do it in another patch.

Reply:
Yes, we can rename EnumerableThetaJoin to EnumerableNestedLoopJoin, and keep EnumerableCorrelate as the LogicalCorrelate's physical implementation, to make things easier.

Reply:
@danny0405 Can we do above in this patch?

Reply:
@hsyuan  i hope to do it in another patch, cause this patch already modified too much files, split it to another patch looks more clear, :), i can do it also in this patch if it take no more burden for reviewing.

Reply:
Our common agreement is Rename EnumerableThetaJoin to NestedLoopJoin, and keep EnumerableCorrelate unchanged. If you do it in another patch, better revert the change for EnumerableCorrelate, don't rename it.

Reply:
The rename is ok, cause it is indeed a NestedLoopJoin and there is on other NestedLoopJoin class in our code, After i rename EnumerableThetaJoin to NestedLoopJoin, these 2 NestedLoopJoin logic are merged. There is no confusing here.

Reply:
I think merging EnumerableCorrelate (which uses EnumerableDefaults#correlateJoin) and EnunmerableThetaJoin (which uses EnumerableDefaults#thetaJoin) into the same EnumerableNestedLoopJoin would result in a single, but complicated class.
I agree with @hsuanyi (and [this comment](https://issues.apache.org/jira/browse/CALCITE-2969?focusedCommentId=16827538&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-16827538)), I think it would be simpler to keep EnumerableCorrelate unchanged (so that it will be clear that it is the Enumerable implementation of Correlate), and rename only EnumerableThetaJoin as EnumerableNestedLoopJoin.
Having said that, I'll agree with the option chosen by the majority.

Reply:
Okey, have made changes:
1. Rename EnumerableThetaJoin to EnumerableNestedLoopJoin
2.Revert change to EnumerableCorrelate

BTW, sub-query.iq does not have better plan for null correlate now :(